mobile/android/android-emulator-services.ts

```diff
@@ -120,29 +147,113 @@ class AndroidEmulatorServices implements Mobile.IEmulatorPlatformServices {
 		return timeout * 1000;
 	}
 
+
+	private getRunningEmulatorId(image: string): IFuture<string> {
+		return ((): string => {
+			var runningEmulators = this.getRunningEmulators().wait();
+			if(runningEmulators.length === 0) {
+				return "";
+			}
+
+			// if we get here, there's at least one running emulator
+			var emulatorId: string;
+
+			_(runningEmulators).each(emulator => {
+				var emulatorName = options.geny ? this.getNameFromGenymotionEmulatorId(emulator).wait() : this.getNameFromNativeEmulatorId(emulator).wait();
+				if(emulatorName === image) {
+					emulatorId = emulator;
+					return false;
+				}
+			});
+
+			return emulatorId;
+		}).future<string>()();
+
+	}
+
+	private getNameFromGenymotionEmulatorId(emulatorId: string): IFuture<string> {
+		return (() => {
+			var modelOutputLines: string = this.$childProcess.execFile(this.$staticConfig.adbFilePath, ["-s", emulatorId, "shell", "getprop", "ro.product.model"]).wait();
+			this.$logger.trace(modelOutputLines);
+			var model = (<string>_.first(modelOutputLines.split(os.EOL))).trim();
+			return model;
+		}).future<string>()();
+	}
+
+	private getNameFromNativeEmulatorId(emulatorId: string): IFuture<string> {
+		var match = emulatorId.match(AndroidEmulatorServices.RUNNING_ANDROID_EMULATOR_REGEX);
+		var portNumber: string;
+		if(match && match[1]) {
+			portNumber = match[1];
+		} else {
+			Future.fromResult("");
```

Comment:
Maybe I do not understand your idea, but I need the future in order to resolve it when client.end is called. Using Future.fromResult is resolving the future, so I cannot resolve it in end event of the client (leads to Future resolved more than once error).